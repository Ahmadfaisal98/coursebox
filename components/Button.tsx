import { MouseEvent } from "react";
import styled from "@emotion/styled";
import { css, SerializedStyles } from "@emotion/react";

import { AppTheme } from "../styles/themes";

export type Color = "primary" | "secondary" | "danger" | "warning";

export type Props = {
  children: string;
  color?: Color;
  onClick: (event: MouseEvent<HTMLButtonElement>) => void;
};

export const getColors = (theme: AppTheme, color?: Color): SerializedStyles => {
  switch (color) {
    case "secondary":
      return css`
        color: ${theme.font.regular};
      `;
    case "primary":
    case "danger":
      return css`
        background: ${theme.components[color]};
        color: ${theme.font.button};
      `;
    case "warning":
      return css`
        background: ${theme.components[color]};
        color: ${theme.font.warning};
      `;
    default:
      return css``;
  }
};

export const Button = styled.button<Props>`
  all: unset;
  display: flex;
  justify-self: center;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  font-size: 1.6rem;
  width: 15rem;
  height: 4rem;
  border-radius: 1rem;
  transition: all 0.4s ease;
  ${({ color, theme }) => getColors(theme, color)};
  ${({ theme }) =>
    `box-shadow: 0.5vmin 0.5vmin 1vmin ${theme.components.shadow1}, -0.5vmin -0.5vmin 1vmin ${theme.components.shadow1}`};
  &:hover {
    opacity: 0.9;
  }
  &:active {
    ${({ theme }) =>
      `box-shadow: 0.5vmin 0.5vmin 1vmin ${theme.components.shadow1} inset, -0.5vmin -0.5vmin 1vmin ${theme.components.shadow1} inset`};
  }
`;

Button.defaultProps = {
  color: "primary",
};
