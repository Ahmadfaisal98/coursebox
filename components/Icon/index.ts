export { Icon, type Props } from "./Icon";
