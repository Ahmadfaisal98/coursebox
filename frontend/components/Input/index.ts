export { Input } from "./Input";
export { Feedback } from "./Feedback";
