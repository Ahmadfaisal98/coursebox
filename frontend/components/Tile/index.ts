export { Tile, type Props } from "./Tile";
export { CenteredTile } from "./CenteredTile";
