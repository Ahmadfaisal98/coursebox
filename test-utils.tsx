import React, { FC, ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@emotion/react";

import { Themes } from "./styles/themes";

const Wrapper: FC = ({ children }) => (
  <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
);

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
