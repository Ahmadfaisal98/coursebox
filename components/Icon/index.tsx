import { FC } from "react";
import styled from "@emotion/styled";

import { Icons } from "./Icon";

export type AvailableIcons = keyof typeof Icons;

type WrapperProps = {
  /** Width and height */
  size: string;
};

export type Props = {
  /** Icon name */
  name: AvailableIcons;
  size: string;
} & WrapperProps &
  React.SVGProps<SVGSVGElement>;

const Wrapper = styled.div<WrapperProps>`
  color: ${({ theme }) => theme.font.regular};
  width: ${({ size }) => size};
  width: ${({ size }) => size};
`;

// https://reactsvgicons.com/search

export const Icon: FC<Props> = ({ name, size = "2rem", ...rest }) => {
  const Icon = Icons[name];
  const sizes = { width: size, height: size };

  return (
    <Wrapper size={size}>
      <Icon {...sizes} {...rest} />
    </Wrapper>
  );
};
