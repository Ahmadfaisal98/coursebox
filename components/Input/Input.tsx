import { FC, ChangeEventHandler, InputHTMLAttributes, ReactChild } from "react";
import styled from "@emotion/styled";

import { Icon, AvailableIcons } from "@/components/Icon";
import { boxShadow } from "@/components/styles";

const StyledInput = styled.input`
  all: unset;
  width: ${({ width }) => width}rem;
  height: ${({ height }) => height}rem;
  border-radius: 1rem;
  font-size: 1.4rem;
  padding: 0 2.6rem 0 1.4rem;
  color: ${({ theme }) => theme.font.regular};
  ${({ theme }) =>
    boxShadow(theme.components.shadow1, theme.components.shadow2, true)}
  &::placeholder {
    color: ${({ theme }) => theme.font.placeholder};
  }
  &:focus {
    ${({ theme }) =>
      boxShadow(theme.components.shadow1, theme.components.shadow2)}
    ~ svg {
      color: ${({ theme }) => theme.font.regular};
      opacity: 1;
    }
  }
`;

const Label = styled.label`
  display: flex;
  justify-content: flex-start;
  flex-direction: column;
  color: ${({ theme }) => theme.font.regular};
  font-size: 1rem;
`;

const StyledIcon = styled(Icon)`
  margin-left: -2.5rem;
  color: ${({ theme }) => theme.font.placeholder};
  opacity: 0.7;
`;

const InputWrapper = styled.div`
  display: flex;
  align-items: center;
`;

const Text = styled.span`
  padding-left: 1.4rem;
`;

export type Props = {
  /** Placeholder  */
  placeholder: string;
  /** onChange handler */
  onChange: ChangeEventHandler<HTMLInputElement>;
  /** Input label */
  label?: string;
  /** Icon */
  icon?: AvailableIcons;
  /** Feedback for input */
  feedback?: ReactChild;
  /** Input height */
  height?: number;
  /** Input width */
  width?: number;
};

export const Input: FC<Props & InputHTMLAttributes<HTMLInputElement>> = ({
  label,
  height = 4,
  width = 20,
  icon,
  feedback,
  ...props
}) => (
  <Label>
    {label && <Text>{label}</Text>}
    <InputWrapper>
      <StyledInput height={height} width={width} {...props} />
      {icon && <StyledIcon name={icon} />}
    </InputWrapper>
    {feedback && <Text>{feedback}</Text>}
  </Label>
);
