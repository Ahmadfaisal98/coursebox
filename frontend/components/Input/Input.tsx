import { FC, ChangeEventHandler, InputHTMLAttributes, ReactChild } from "react";
import styled from "@emotion/styled";

import { Icon, AvailableIcons } from "@/components/Icon";
import { boxShadow } from "@/components/styles";

const StyledInput = styled.input`
  all: unset;
  width: 100%;
  height: 100%;
  border-radius: 1rem;
  font-size: 1.4rem;
  padding: 0 2.6rem 0 1.4rem;
  color: ${({ theme }) => theme.font.regular};
  ${({ theme }) =>
    boxShadow(theme.components.shadow1, theme.components.shadow2, true)}
  &::placeholder {
    color: ${({ theme }) => theme.font.placeholder};
  }
  &:focus {
    ${({ theme }) =>
      boxShadow(theme.components.shadow1, theme.components.shadow2)}
    ~ svg {
      color: ${({ theme }) => theme.font.regular};
      opacity: 1;
    }
  }
`;

type LabelProps = {
  /** Input height */
  height?: number;
  /** Input width */
  width?: number;
};

const Label = styled.label<LabelProps>`
  display: flex;
  justify-content: flex-start;
  flex-direction: column;
  width: ${({ width }) => width}rem;
  height: ${({ height }) => height}rem;
  color: ${({ theme }) => theme.font.regular};
  font-size: 1rem;
`;

const StyledIcon = styled(Icon)`
  position: absolute;
  right: 0.3rem;
  color: ${({ theme }) => theme.font.placeholder};
  opacity: 0.7;
`;

const InputWrapper = styled.div`
  display: flex;
  position: relative;
  align-items: center;
  width: 100%;
  height: 100%;
`;

const Text = styled.span`
  padding-left: 1.4rem;
`;

export type Props = {
  /** Placeholder  */
  placeholder: string;
  /** onChange handler */
  onChange: ChangeEventHandler<HTMLInputElement>;
  /** Input label */
  label?: string;
  /** Icon */
  icon?: AvailableIcons;
  /** Feedback for input */
  feedback?: ReactChild;
} & LabelProps;

export const Input: FC<Props & InputHTMLAttributes<HTMLInputElement>> = ({
  label,
  height = 7,
  width = 20,
  icon,
  feedback,
  className,
  ...props
}) => (
  <Label height={height} width={width} className={className}>
    {label && <Text>{label}</Text>}
    <InputWrapper>
      <StyledInput {...props} />
      {icon && <StyledIcon name={icon} />}
    </InputWrapper>
    {feedback && <Text>{feedback}</Text>}
  </Label>
);
