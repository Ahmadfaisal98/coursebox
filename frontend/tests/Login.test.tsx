import userEvent from "@testing-library/user-event";

import { Feedback } from "@/components/Input";
import { render, screen, act } from "@/test-utils";

import Login from "@/pages/login";

describe("Login page", () => {
  it("Render check", () => {
    const { container } = render(<Login />);
    expect(container).toMatchSnapshot();
  });
});
