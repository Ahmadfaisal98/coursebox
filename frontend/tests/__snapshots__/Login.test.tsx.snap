// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page Render check 1`] = `
<div>
  <form>
    <div
      class="css-qbqlvy"
    >
      <section
        class="css-6q0t2w"
      >
        <h2>
          Login
        </h2>
        <label
          class="css-bhzw8w"
          height="7"
          width="20"
        >
          <span
            class="css-1grri41"
          >
            Identifier
          </span>
          <div
            class="css-52ovgc"
          >
            <input
              class="css-1bj11r8"
              placeholder="username or email"
            />
          </div>
          <span
            class="css-1reg724"
          />
        </label>
        <label
          class="css-bhzw8w"
          height="7"
          width="20"
        >
          <span
            class="css-1grri41"
          >
            Password
          </span>
          <div
            class="css-52ovgc"
          >
            <input
              class="css-1bj11r8"
              placeholder="password"
              type="password"
            />
          </div>
          <span
            class="css-1reg724"
          />
        </label>
        <button
          class="css-1tkemmz"
          color="primary"
          type="submit"
        >
          Sign In
        </button>
        <h3>
          <a
            class="css-pm0vza"
            href="/registration"
          >
            Create account
          </a>
        </h3>
      </section>
    </div>
  </form>
</div>
`;
